projects/driver_controls/src/gpio_expander.c

```diff
@@ -0,0 +1,168 @@
+
+    if (s_interrupts[current_pin].callback != NULL) {
+      s_interrupts[current_pin].callback(current_pin, s_interrupts[current_pin].context);
+    }
+
+    s_register[MCP23008_INTF] &= ~(1 << current_pin);
+  }
+
+  // Clear the interrupt by reading the port registers
+  prv_update_registers();
+}
+
+// Set the value of a specific bit in a register
+static void prv_set_bit(uint8_t *data, GPIOExpanderPin pin, bool bit) {
+  if (bit) {
+    *data |= (1 << pin);
+  } else {
+    *data &= ~(1 << pin);
+  }
+}
+
+StatusCode gpio_expander_init(GPIOAddress address, I2CPort i2c_port) {
+  // Set the static variables
+  s_address = address;
+  s_i2c_port = i2c_port;
+
+  // Configure the pin to receive interrupts from the MCP23008
+  GPIOSettings gpio_settings = { .direction = GPIO_DIR_IN, .alt_function = GPIO_ALTFN_NONE };
+  InterruptSettings it_settings = { INTERRUPT_TYPE_INTERRUPT, INTERRUPT_PRIORITY_NORMAL };
+
+  gpio_init_pin(&address, &gpio_settings);
+  gpio_it_register_interrupt(&address, &it_settings, INTERRUPT_EDGE_FALLING,
+                              prv_interrupt_handler, NULL);
+
+  // Initialize the interrupt callbacks to NULL
+  for (uint8_t i = 0; i < NUM_GPIO_EXPANDER_PIN; i++) {
+    s_interrupts[i].callback = NULL;
+  }
+
+  // Use Sequential write to set each register to the default value
+  i2c_write_reg(s_i2c_port, MCP23008_ADDRESS, MCP23008_IODIR, &s_register[MCP23008_IODIR], 11);
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode gpio_expander_init_pin(GPIOExpanderPin pin, GPIOSettings *settings) {
+  StatusCode valid = prv_pin_is_valid(pin);
+  status_ok_or_return(valid);
+
+  if (pin >= NUM_GPIO_EXPANDER_PIN) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // Set the direction of the data I/O
+  prv_set_bit(&s_register[MCP23008_IODIR], pin, (settings->direction == GPIO_DIR_IN));
+  i2c_write_reg(s_i2c_port, MCP23008_ADDRESS, MCP23008_IODIR, &s_register[MCP23008_IODIR], 1);
+
+  if (settings->direction == GPIO_DIR_OUT) {
+    // Set the output state if the pin is an output
+    prv_set_bit(&s_register[MCP23008_GPIO], pin, (settings->state == GPIO_STATE_HIGH));
+    i2c_write_reg(s_i2c_port, MCP23008_ADDRESS, MCP23008_GPIO, &s_register[MCP23008_GPIO], 1);
+  } else {
+    // Enable interrupts if the pin is an input
+    prv_set_bit(&s_register[MCP23008_GPINTEN], pin, true);
+    i2c_write_reg(s_i2c_port, MCP23008_ADDRESS, MCP23008_GPINTEN, &s_register[MCP23008_GPINTEN], 1);
+
+    // Configure so that any change in value triggers an interrupt
+    prv_set_bit(&s_register[MCP23008_INTCON], pin, false);
+    i2c_write_reg(s_i2c_port, MCP23008_ADDRESS, MCP23008_INTCON, &s_register[MCP23008_INTCON], 1);
+  }
+
+  return STATUS_CODE_OK;
+}
+
+bool gpio_expander_get_state(GPIOExpanderPin pin) {
+  StatusCode valid = prv_pin_is_valid(pin);
+  status_ok_or_return(valid);
+
+  // Update our GPIO register value and read the pin state
+  prv_update_registers();
```

Comment:
My point is that when you're running your callback, you already read the state information. Why not just pass it in as part of the callback instead of forcing them to run another function that needs to read i2c again?